backend/src/main/java/fk/prof/backend/model/assignment/AggregationWindowPlanner.java

```diff
@@ -0,0 +1,203 @@
+    this.simultaneousWorkAssignmentCounter = Preconditions.checkNotNull(simultaneousWorkAssignmentCounter);
+    this.aggregationWindowDurationInMins = aggregationWindowDurationInMins;
+    this.workProfileRefreshBufferInSecs = workProfileRefreshBufferInSecs;
+
+    this.aggregationWindowScheduleTimer = Future.future();
+    getWorkForNextAggregationWindow(currentAggregationWindowIndex + 1).setHandler(ar -> {
+      aggregationWindowSwitcher();
+
+      // From vertx docs:
+      // Keep in mind that the timer will fire on a periodic basis.
+      // If your periodic treatment takes a long amount of time to proceed, your timer events could run continuously or even worse : stack up.
+      // NOTE: The above is a fringe scenario since aggregation window duration is going to be in excess of 20 minutes
+      // Still, there is a way to detect if this build-up happens. If aggregation window switch event happens before work profile is fetched, we publish a metric
+      // If /leader/work API latency is within bounds but this metric is high, this implies a build-up of aggregation window events
+      long periodicTimerId = vertx.setPeriodic(aggregationWindowDurationInMins * 60 * MILLIS_IN_SEC,
+          timerId -> aggregationWindowSwitcher());
+      this.aggregationWindowScheduleTimer.complete(periodicTimerId);
+    });
+  }
+
+  /**
+   * This expires current aggregation window and cancels scheduling of upcoming aggregation windows
+   * To be called when leader de-associates relevant process group from the backend
+   */
+  public void close() {
+    aggregationWindowScheduleTimer.setHandler(ar -> {
+      if(ar.succeeded()) {
+        vertx.cancelTimer(ar.result());
+      }
+    });
+    expireCurrentAggregationWindow();
+  }
+
+  /**
+   * This method will be called before start of every aggregation window
+   * There should be sufficient buffer to allow completion of this method before the next aggregation window starts
+   * Not adding any guarantees here, but a lead of few minutes for this method's execution should ensure that the request to get work should complete in time for next aggregation window
+   */
+  private Future<Void> getWorkForNextAggregationWindow(int aggregationWindowIndex) {
+    latestWorkProfile = null;
+    Future<Void> result = Future.future();
+    this.workForBackendRequestor.apply(processGroup).setHandler(ar -> {
+      relevantAggregationWindowIndexForWorkProfile = aggregationWindowIndex;
+      if(ar.failed()) {
+        //Cannot fetch work from leader, so chill out and let this aggregation window go by
+        //TODO: Metric to indicate failure to fetch work for this process group from leader
+        logger.error("Error fetching work from leader for process_group=" + RecorderProtoUtil.processGroupCompactRepr(processGroup) + ", error=" + ar.cause().getMessage());
+        result.fail(ar.cause());
+      } else {
+        latestWorkProfile = ar.result();
```

Comment:
Fixed in commit https://github.com/Flipkart/fk-prof/commit/eaa57eb6a8bb821fd8543e80300e2965bbafcb28